SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc

```diff
@@ -2607,3 +2610,130 @@ void SiPixelDigitizerAlgorithm::lateSignalReweight(const PixelGeomDetUnit* pixde
   digis.clear();
   digis = New_digis;
 }
+
+/******************************************************************/
+
+void SiPixelDigitizerAlgorithm::lateSignalReweight(const PixelGeomDetUnit* pixdet,
+                                                   std::vector<PixelDigi>& digis,
+                                                   std::vector<PixelSimHitExtraInfoLite>& newClass_Sim_extra,
+                                                   const TrackerTopology* tTopo,
+                                                   CLHEP::HepRandomEngine* engine) {
+  // Function to apply the Charge Reweighting on top of digi in case of PU from mixing library
+  // for time dependent MC
+  std::vector<PixelDigi> New_digis;
+  uint32_t detID = pixdet->geographicalId().rawId();
+
+  if (UseReweighting) {
+    LogError("PixelDigitizer ") << " ********************************  \n";
+    LogError("PixelDigitizer ") << " ********************************  \n";
+    LogError("PixelDigitizer ") << " *****  INCONSISTENCY !!!   *****  \n";
+    LogError("PixelDigitizer ")
+        << " applyLateReweighting_ and UseReweighting can not be true at the same time for PU ! \n";
+    LogError("PixelDigitizer ") << " ---> DO NOT APPLY CHARGE REWEIGHTING TWICE !!! \n";
+    LogError("PixelDigitizer ") << " ******************************** \n";
+    LogError("PixelDigitizer ") << " ******************************** \n";
+    return;
+  }
+
+  float thePixelThresholdInE = 0.;
+  if (theNoiseInElectrons > 0.) {
+    if (pixdet->type().isTrackerPixel() && pixdet->type().isBarrel()) {  // Barrel modules
+      int lay = tTopo->layer(detID);
+      if (addThresholdSmearing) {
+        if (pixdet->subDetector() == GeomDetEnumerators::SubDetector::PixelBarrel ||
+            pixdet->subDetector() == GeomDetEnumerators::SubDetector::P1PXB) {
+          if (lay == 1) {
+            thePixelThresholdInE = CLHEP::RandGaussQ::shoot(
+                engine, theThresholdInE_BPix_L1, theThresholdSmearing_BPix_L1);  // gaussian smearing
+          } else if (lay == 2) {
+            thePixelThresholdInE = CLHEP::RandGaussQ::shoot(
+                engine, theThresholdInE_BPix_L2, theThresholdSmearing_BPix_L2);  // gaussian smearing
+          } else {
+            thePixelThresholdInE =
+                CLHEP::RandGaussQ::shoot(engine, theThresholdInE_BPix, theThresholdSmearing_BPix);  // gaussian smearing
+          }
```

Comment:
@ferencek , can theoretically a threshold become negative after smearing?

Reply:
Note that this is not a feature specific to the newly added code but is appearing in several other places in the code. In theory the threshold could become negative but looking at parameter values in https://github.com/cms-sw/cmssw/blob/CMSSW_14_2_0/SimGeneral/MixingModule/python/SiPixelSimParameters_cfi.py#L65-L73, this would have to be more than a 10-sigma fluctuation. As far as I see, the only implication of such a negative threshold would be that any positive signal or noise would simply pass the threshold which is in principle ok and would not lead to code crashing. Not sure if @carolinecollard has anything to add. So I would say this does not require a fix.